src/rust/cryptography-x509-validation/src/lib.rs

```diff
@@ -11,6 +11,342 @@ pub mod policy;
+            }
+        }
+
+        if let Some(new_constraints) = new_constraints {
+            self.name_constraints.push(new_constraints);
+        }
+
+        Ok(())
+    }
+}
+
+pub type Chain<'c> = Vec<Certificate<'c>>;
+
+pub fn verify<'a, 'chain, B: CryptoOps>(
+    leaf: &'a Certificate<'chain>,
+    intermediates: impl IntoIterator<Item = Certificate<'chain>>,
+    policy: &Policy<'_, B>,
+    store: &'a Store<'chain>,
+) -> Result<Chain<'chain>, ValidationError> {
+    let builder = ChainBuilder::new(intermediates.into_iter().collect(), policy, store);
+
+    builder.build_chain(leaf)
+}
+
+struct ChainBuilder<'a, 'chain, B: CryptoOps> {
+    intermediates: HashSet<Certificate<'chain>>,
+    policy: &'a Policy<'a, B>,
+    store: &'a Store<'chain>,
+}
+
+// When applying a name constraint, we need to distinguish between a few different scenarios:
+// * `Applied(true)`: The name constraint is the same type as the SAN and matches.
+// * `Applied(false)`: The name constraint is the same type as the SAN and does not match.
+// * `Skipped`: The name constraint is a different type to the SAN.
+enum ApplyNameConstraintStatus {
+    Applied(bool),
+    Skipped,
+}
+
+impl ApplyNameConstraintStatus {
+    fn is_applied(&self) -> bool {
+        matches!(self, Applied(_))
+    }
+
+    fn is_match(&self) -> bool {
+        matches!(self, Applied(true))
+    }
+}
+
+impl<'a, 'chain, B: CryptoOps> ChainBuilder<'a, 'chain, B> {
+    fn new(
+        intermediates: HashSet<Certificate<'chain>>,
+        policy: &'a Policy<'a, B>,
+        store: &'a Store<'chain>,
+    ) -> Self {
+        Self {
+            intermediates,
+            policy,
+            store,
+        }
+    }
+
+    fn potential_issuers(
+        &'a self,
+        cert: &'a Certificate<'chain>,
+    ) -> impl Iterator<Item = &'a Certificate<'chain>> + '_ {
+        // TODO: Optimizations:
+        // * Use a backing structure that allows us to search by name
+        //   rather than doing a linear scan
+        // * Search by AKI and other identifiers?
+        self.store
+            .iter()
+            .chain(self.intermediates.iter())
+            .filter(|&candidate| candidate.subject() == cert.issuer())
+    }
+
+    fn build_chain_inner(
+        &self,
+        working_cert: &'a Certificate<'chain>,
+        current_depth: u8,
+        working_cert_extensions: &'a Extensions<'chain>,
+        accumulated_constraints: &'a mut AccumulatedNameConstraints<'chain>,
+    ) -> Result<Chain<'chain>, ValidationError> {
+        // Per RFC 5280: Name constraints are not applied
+        // to subjects in self-issued certificates, *unless* the
+        // certificate is the final (i.e., leaf) certificate in the path.
+        //
+        // See: RFC 5280 4.2.1.10
+        let skip_name_constraints = cert_is_self_issued(working_cert) && current_depth != 0;
+        if !skip_name_constraints {
+            accumulated_constraints.accumulate_san(working_cert_extensions)?;
```

Comment:
So I think this is wrong for the following situation:

```
EE +-> Intermediate 1 (SAN) -> No root to chain to
   |-> Intermediate 2 (no SAN) -> Root (NC that doesn't match Intermediate 1 SAN)
```

Reply:
Added to limbo here: https://trailofbits.github.io/x509-limbo/testcases/rfc5280/#rfc5280ncnc-forbids-alternate-chain-san

Working on this now.